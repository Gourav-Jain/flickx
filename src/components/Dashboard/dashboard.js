import React from 'react';
import { Input, Button, Segment, Popup, Message } from 'semantic-ui-react';
import { useDashboardContainer } from './dashboard.container'
import { Movies } from '../Movies/movies'
import { Footer } from '../Footer/footer'

export const Dashboard = () => {

    const { showLoader,
        error,
        errorMessage,
        handleChange,
        handleAddToFav,
        handleRemoveAddToFave,
        handleAddToWatchLater,
        handleRemoveWatchLater,
        result,
        favourites,
        watchLater } = useDashboardContainer();

    return (
        <>
            <Segment clearing>

                <div>
                    <Popup content="Favorites" trigger={
                        <Button
                            floated='right'
                            color='yellow'
                            content=''
                            icon='favorite'
                            label={{ basic: true, color: 'black', pointing: 'left', content: favourites.length }}
                        />}
                    />
                    <Popup content="Watch Later" trigger={<Button
                        color='blue'
                        content=''
                        icon='time'
                        floated='right'
                        label={{
                            as: 'a',
                            basic: true,
                            color: 'black',
                            pointing: 'left',
                            content: watchLater.length,
                        }}
                    />}
                    />
                </div>
                <Input floated='right' loading={showLoader} size='large' icon='search' placeholder='Search...' onChange={handleChange} />
            </Segment>
            {
                error && errorMessage ? <Message error={error}>
                    <p>{errorMessage}</p>
                </Message> : <Movies result={result} addToFavo={handleAddToFav} addToWatchLater={handleAddToWatchLater} showLoader={showLoader} />
            }


            <Footer favourites={favourites} removeFavourite={handleRemoveAddToFave} watchLater={watchLater} removeWatchLater={handleRemoveWatchLater} />
        </>
    )
}

export default Dashboard;