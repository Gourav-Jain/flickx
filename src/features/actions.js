/**
 * Action Types
 */
export const SEARCH_MOVIES_REQS = "SEARCH_MOVIES_REQS";
export const SEARCH_MOVIES_RESP = "SEARCH_MOVIES_RESP";
export const SEARCH_MOVIES_FAIL = "SEARCH_MOVIES_FAIL";

export const ADD_TO_FAVOURITE = "ADD_TO_FAVOURITE";
export const ADD_TO_WATCH_LATER = "ADD_TO_WATCH_LATER";
export const REMOVE_FROM_FAVOURITE = "REMOVE_FROM_FAVOURITE";
export const REMOVE_FROM_WATCH_LATER = "REMOVE_FROM_WATCH_LATER";

/**
 * Action Creator
 */
export const setSearchMoviesRequest = payload => ({ type: SEARCH_MOVIES_REQS, payload });
export const setSearchMoviesResponse = payload => ({ type: SEARCH_MOVIES_RESP, payload });
export const setSearchMoviesFailer = payload => ({ type: SEARCH_MOVIES_FAIL, payload });

export const setAddToFavourites = payload => ({ type: ADD_TO_FAVOURITE, payload });
export const setAddToWatchLater = payload => ({ type: ADD_TO_WATCH_LATER, payload });
export const setRemoveFromFavourite = payload => ({ type: REMOVE_FROM_FAVOURITE, payload });
export const setRemoveFromWatchLater = payload => ({ type: REMOVE_FROM_WATCH_LATER, payload });