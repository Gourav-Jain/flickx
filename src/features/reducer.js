import {
    SEARCH_MOVIES_REQS,
    SEARCH_MOVIES_RESP,
    SEARCH_MOVIES_FAIL,
    ADD_TO_FAVOURITE,
    ADD_TO_WATCH_LATER,
    REMOVE_FROM_FAVOURITE,
    REMOVE_FROM_WATCH_LATER
} from "./actions";
import initialState from '../store/initialState';

export const reducer = (state = initialState, action) => {
    switch (action.type) {
        case SEARCH_MOVIES_REQS:
            return {
                ...state,
                ...action.payload
            }
        case SEARCH_MOVIES_RESP:
            const { favourites, watchLater } = state || [];
            let { searchResult } = action.payload;
            if (favourites) {
                favourites.forEach(el => {
                    const idx = searchResult.result.findIndex(i => i.id === el.id);
                    if (idx !== -1) {
                        searchResult.result[searchResult.result.findIndex(i => i.id === el.id)].fav = true;
                    }
                });
            }

            if (watchLater) {
                watchLater.forEach(el => {
                    const idx = searchResult.result.findIndex(i => i.id === el.id);
                    if (idx !== -1) {
                        searchResult.result[searchResult.result.findIndex(i => i.id === el.id)].wl = true;
                    }
                });
            }
            const newState = { ...state, ...action.payload }

            return newState;
        case SEARCH_MOVIES_FAIL:
            return { ...state, ...action.payload };
        case ADD_TO_FAVOURITE:
            return action.payload;
        case ADD_TO_WATCH_LATER:
            return action.payload;
        case REMOVE_FROM_FAVOURITE:
            return action.payload;
        case REMOVE_FROM_WATCH_LATER:
            return action.payload;

        default:
            return state;
    }
};