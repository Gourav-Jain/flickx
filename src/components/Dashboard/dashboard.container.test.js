import { renderHook, act } from "@testing-library/react-hooks";

import { useDashboardContainer } from "./dashboard.container";
import * as action from "../../features/actions";
import {
    useSelectorHook, getSearchMovieApiUrl, getPopularMovieApiUrl,
    API_KEY, API_SERACH_MOVIE_URL, API_POPULAR_MOVIE_URL
} from '../../utils'


jest.mock("react-redux", () => ({
    useDispatch: jest.fn,
    useSelector: () => ({
        searchMovies: {}
    })
}));

const mockPush = jest.fn();
const mockUseDispatch = jest.fn(mockPush);
const handleChange = jest.fn();
describe("useDashboardContainer()", () => {

    beforeEach(jest.clearAllMocks);

    const render = () => {
        const { result } = renderHook(useDashboardContainer);

        return result.current;
    };

    describe("@returns property", () => {

        test("should return data from store", () => {
            const { showLoader } = render();

            expect(showLoader).toMatchSnapshot();
        });

        describe("should return showLoader", () => {
            test("as false on initial render", () => {
                const { showLoader } = render();

                expect(showLoader).toBe(false);
            });
        });
    });
});
