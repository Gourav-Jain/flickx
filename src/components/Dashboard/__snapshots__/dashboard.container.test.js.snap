// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useDashboardContainer() @returns property should return data from store 1`] = `false`;
